### PR TITLE
Improvements to when we get_store_cost

### PR DESCRIPTION
**Pull request by joshuef**
Adds some caching to store cost for the client.

atop #624

### CODE REVIEW

File: sn_client/src/api.rs [hunk 1]
```diff
@@ -419,16 +431,31 @@ impl Client {
...
+            .as_nano();
+
+        if average_cost {
+            if let Some(current_cost) = self.network_store_cost {
+                self.network_store_cost = Some((current_cost + cost) / 2);
+            } else {
+                self.network_store_cost = Some(cost);
+            }
+        } else {
+            self.network_store_cost = Some(cost);
```
Comment (maqi) [inline sn_client/src/api.rs]: if not averaging, shall we consider only update when the incoming cost is higher ?

File: sn_client/src/wallet.rs [hunk 2]
```diff
@@ -75,13 +81,13 @@ impl WalletClient {
     }
 
     /// Get storecost from the network
-    /// this uses the largest DBC as the payment section
-    pub async fn get_store_cost(&mut self) -> Result<Token> {
+    /// Stores this value as the new baseline at the client
+    pub async fn set_store_cost_from_random_address(&mut self) -> Result<Token> {
         // For now we simply just use the largest DBC we have...
```
Comment (maqi) [inline sn_client/src/wallet.rs]: the comment needs to be updated

File: sn_client/src/wallet.rs [hunk 3]
```diff
@@ -90,6 +96,9 @@ impl WalletClient {
     ///
     /// Returns (Proofs and an Option around Storage Cost), storage cost is _per record_, and only returned if required for this operation
     ///
+    /// `pay_store_cost_at` is the cost to pay for the storage, if None, the cost will be retrieved from the network.
+    /// Normally this will be a cast sampled at client init. It may be this has been user set and passed into the Client.
```
Comment (maqi) [inline sn_client/src/wallet.rs]: cast -> cost

File: sn_networking/src/lib.rs [hunk 4]
```diff
@@ -988,16 +989,31 @@ impl Network {
...
     all_costs.sort();
 
-    let token_fee = *all_costs
-        .get(target_cost_index)
-        .ok_or(Error::NotEnoughCostQuotes)?;
+    let token_fee = if any_cost_will_do {
+        let first_try = all_costs
+            .get(target_cost_index)
+            .ok_or(Error::NotEnoughCostQuotes);
+        // first try our ideal, otherwise just use the last one
```
Comment (maqi) [inline sn_networking/src/lib.rs]: Prefer the value from first_try, otherwise just use the value from the last.

File: sn_networking/src/lib.rs [hunk 5]
```diff
@@ -1081,6 +1083,35 @@ mod tests {
...
 
+        Ok(())
+    }
+    #[test]
+    fn test_get_any_fee_from_store_cost_quotes() -> eyre::Result<()> {
+        // for a vec of different costs of CLOUSE_GROUP size
+        // ensure we return the CLOSE_GROUP / 2 indexed price
+        let mut costs = vec![];
+        for i in 0..(CLOSE_GROUP_SIZE / 2) - 1 {
+            println!("price: {}", i);
```
Comment (maqi) [inline sn_networking/src/lib.rs]: the println can now be removed ?

File: sn_client/src/api.rs (hunk 1 above)
Comment (joshuef) [reply sn_client/src/api.rs]: I don't think so? The price could reduce as much as go up?

Comment (maqi) [reply sn_client/src/api.rs]: well, it's bit tricky here as I understand.
This storage_cost will be used for a payment targeting any part of the network.
If talking about a snapshot of the whole network's storage cost distribution at specific time, it shall always be the highest one to prefer, to secure the cost we pay can be accepted by any part of the network.

However, as you said, if talking about the storage cost along the time, it does can go down as well.

In general, I feel picking the highest one is the safest way to ensure the payment can be accepted across the network, within the current mechanism of how storage cost got calculated by client and validated by network nodes on receiving a payment.

Comment (joshuef) [reply sn_client/src/api.rs]: Ahhh, yeh, i totally agree with what you're saying. I thought that's what we were doing here but it's not really.

We're averaging, which means we won't pay to a part of the network. So we actually should not avg, but just increase if higher.

**Comment by joshuef** [comment]
atop #624

File: sn_client/src/api.rs [hunk 6]
```diff
@@ -144,13 +151,13 @@ impl Client {
...
+                    debug!("No ClientEvent activity in the past {timeout:?}, performing a random get_store_cost query to target: {random_target:?}");
                     if is_connected {
-                        println!("The client experienced inactivity in the past {timeout:?}.");
+                        info!("The client was inactive for {timeout:?}.");
                     } else {
-                        println!("The client still does not know enough network nodes.");
+                        info!("The client still does not know enough network nodes.");
                     }
-                    let _ = client.get_chunk(random_target).await;
+
```
Comment (RolandSherwin) [inline sn_client/src/api.rs:158]: I think a call to fetch a random `get_store_cost` is missing here.

Comment (joshuef) [reply sn_client/src/api.rs:158]: I removed that intentionally as it doesn't seem to help much at all, and confuses the logs

**Review by RolandSherwin — APPROVED** [review]

File: sn_client/src/api.rs [hunk 7]
```diff
@@ -144,13 +151,13 @@ impl Client {
                 }
                 Ok(ClientEvent::InactiveClient(timeout)) => {
                     let random_target = ChunkAddress::new(XorName::random(&mut rng));
-                    debug!("No ClientEvent activity in the past {timeout:?}, performing a random get_chunk query to target: {random_target:?}");
+                    debug!("No ClientEvent activity in the past {timeout:?}, performing a random get_store_cost query to target: {random_target:?}");
```
Comment (RolandSherwin) [inline sn_client/src/api.rs]: This log line has to be updated as we do not perform random queries on inactivity anymore.

File: sn_client/src/wallet.rs [hunk 8]
```diff
@@ -43,6 +45,16 @@ impl WalletClient {
         }
     }
 
+    /// Get any known store cost estimate
+    pub fn store_cost(&self) -> Token {
+        Token::from_nano(self.client.network_store_cost)
+    }
+
+    /// Do we have any uncomfirmed transactions?
```
Comment (RolandSherwin) [inline sn_client/src/wallet.rs]:
```suggestion
    /// Do we have any unconfirmed transactions?
```

File: sn_client/src/wallet.rs [hunk 9]
```diff
@@ -90,6 +101,9 @@ impl WalletClient {
     ///
     /// Returns (Proofs and an Option around Storage Cost), storage cost is _per record_, and only returned if required for this operation
     ///
+    /// `pay_store_cost_at` is the cost to pay for the storage, if None, the cost will be retrieved from the network.
+    /// Normally this will be a cost sampled at client init. It may be this has been user set and passed into the Client.
+    ///
```
Comment (RolandSherwin) [inline sn_client/src/wallet.rs]: I think this argument is not in use anymore?

**Pull request merged**